### ISSUE TITLE
two Integer division bug

### ISSUE THREAD

**Issue by lmtoo** [thymeleaf/thymeleaf#499]
I'm writing a pager fragment,but i occur a division problom between two integers.

> ```
> <ul class="pagination" th:with="batchSize=10,batch=${page.number div batchSize},batchCount=${page.totalPages%batchSize==0?page.totalPages/batchSize:(page.totalPages/batchSize+1)},first=${batch==0},last=${batch==batchCount}">
>     <li th:class="${first?'disabled':''}">
>         <a th:href="${url}+'?page='+${(batch-1)*batchSize}" aria-label="Previous">
>             <span aria-hidden="true">&laquo;</span>
>         </a>
>     </li>
>     <li >
>         <a th:text="${page.number}+'/'+${batchSize}+'='+${page.number/batchSize}"></a>
>     </li>
> ```

```
    <th:block th:each="i: ${#numbers.sequence(batch*batchSize,(batch+1)*batchSize-1)}">
        <li th:if="${page.totalPages gt i}" th:class="${i==page.number?'active':''}">
            <a th:href="${url}+'?page='+${i}" th:text="${i+1}"/>
        </li>
    </th:block>

    <li th:class="${last?'disabled':''}">
        <a th:href="${url}+'?page='+${(batch+1)*batchSize}" aria-label="Next">
            <span aria-hidden="true">&raquo;</span>
        </a>
    </li>
</ul>
```

in the code `<li >
        <a th:text="${page.number}+'/'+${batchSize}+'='+${page.number/batchSize}"></a>
    </li>`

page.number division batchSize,when page.number less than 5 is right and the result is 0,but when page.number is 6~9,the result is 1,that is confuse me......

**Comment by danielfernandez** [comment]
Confirmed and fixed. The reason for this is that `th:with`, when it processes `batchSize`, creates a _number token_ which contains `10` as a `BigDecimal`. Until this fix, all _number tokens_ used `BigDecimal` for their values.

The problem with this is that `10` is converted (as it should be) into a `BigDecimal` with `scale == 0`, but both OGNL and SpringEL take the decision that the result of a division between whatever (e.g. an `int`) and a `BigDecimal` with `scale == 0` must necessarily have `scale == 0` too (which is OK), and should be rounded using `RoundingMode.HALF_EVEN`, which IMHO is wrong (it should be treated as a special case, and applied `RoundingMode.FLOOR`).

See https://github.com/spring-projects/spring-framework/blob/7d4d3db0ddd90fe70fd305b27410124301da1b45/spring-expression/src/main/java/org/springframework/expression/spel/ast/OpDivide.java#L55-L60 in order to understand what SpringEL does (OGNL's code is equivalent to that).

So in the end I took the decision to modify how the values of the _number tokens_ themselves are stored, so that any `BigDecimal` value with `scale == 0` (because it is represented without a decimal point, like your `batchSize=10`) will actually be turned into a `BigInteger`. This changes the way the division operation is applied both in OGNL and in SpringEL and makes your scenario work as expected.
